File: src/frontend/types/estimate.ts
export type EstimateStatus =
  | "draft"
  | "pending_approval"
  | "approved"
  | "sent"
  | "accepted"
  | "rejected"
  | "expired";

export interface EstimateLineItem {
  id: string;
  description: string;
  quantity: number;
  unitPrice: number;
  total: number;
}

export interface Estimate {
  id: string;
  customerId: string;
  customerName: string;
  leadId?: string;
  status: EstimateStatus;
  serviceType: string;
  propertyAddress: string;
  propertySize?: number;
  lineItems: EstimateLineItem[];
  subtotal: number;
  tax: number;
  total: number;
  validUntil: Date;
  createdAt: Date;
  updatedAt: Date;
  createdBy: string; // "ai" or user name
  approvedBy?: string;
  approvedAt?: Date;
  sentAt?: Date;
  respondedAt?: Date;
  notes?: string;
  internalNotes?: string;
}

export interface EstimateTemplate {
  id: string;
  name: string;
  serviceType: string;
  description: string;
  defaultLineItems: Omit<EstimateLineItem, "id">[];
  pricingRules?: {
    minPropertySize?: number;
    maxPropertySize?: number;
    pricePerSquareFoot?: number;
    basePrice?: number;
  };
}

File: src/backend/models/estimate.py
import uuid
from datetime import datetime, date
from sqlalchemy import String, Integer, Numeric, Text, DateTime, Date, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from database import Base


class EstimateLineItem(Base):
    __tablename__ = "estimate_line_items"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    estimate_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("estimates.id", ondelete="CASCADE"), nullable=False)
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(10, 2), default=1)
    unit_price: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    total: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)

    estimate: Mapped["Estimate"] = relationship("Estimate", back_populates="line_items")


class Estimate(Base):
    __tablename__ = "estimates"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    customer_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("customers.id", ondelete="CASCADE"), nullable=False)
    lead_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("leads.id", ondelete="SET NULL"), nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="draft")
    service_type: Mapped[str] = mapped_column(String(100), nullable=False)
    property_address: Mapped[str | None] = mapped_column(String(255), nullable=True)
    property_size: Mapped[int | None] = mapped_column(Integer, nullable=True)
    subtotal: Mapped[float] = mapped_column(Numeric(10, 2), default=0.00)
    tax: Mapped[float] = mapped_column(Numeric(10, 2), default=0.00)
    total: Mapped[float] = mapped_column(Numeric(10, 2), default=0.00)
    valid_until: Mapped[date | None] = mapped_column(Date, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    internal_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_by: Mapped[str] = mapped_column(String(255), nullable=False)
    approved_by: Mapped[str | None] = mapped_column(String(255), nullable=True)
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    responded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    customer: Mapped["Customer"] = relationship("Customer", back_populates="estimates")
    line_items: Mapped[list["EstimateLineItem"]] = relationship("EstimateLineItem", back_populates="estimate", order_by="EstimateLineItem.sort_order", cascade="all, delete-orphan")
    tasks: Mapped[list["Task"]] = relationship("Task", back_populates="estimate", lazy="select")
    activities: Mapped[list["Activity"]] = relationship("Activity", back_populates="estimate", lazy="select")
